### PR TITLE
otelzap: Implement methods on encoder

### PR DESCRIPTION
**Pull request by khushijain21**
Part of https://github.com/open-telemetry/opentelemetry-go-contrib/issues/5191

Pre-work https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5279

### CODE REVIEW

File: bridges/otelzap/encoder_test.go [hunk 1]
```diff
@@ -187,6 +220,52 @@ func TestArrayEncoder(t *testing.T) {
...
+	}))
+}
+
+type turduckens int
+
+func (t turduckens) MarshalLogArray(enc zapcore.ArrayEncoder) error {
+	var err error
+	tur := turducken{}
+	for i := 0; i < int(t); i++ {
+		err = multierr.Append(err, enc.AppendObject(tur))
```
Comment (MrAlias) [inline bridges/otelzap/encoder_test.go:246]: Can we use `errors.Join` instead of adding this import?

File: bridges/otelzap/encoder_test.go [hunk 2]
```diff
@@ -187,6 +220,52 @@ func TestArrayEncoder(t *testing.T) {
 	}
 }
 
+type turducken struct{}
```
Comment (MrAlias) [inline bridges/otelzap/encoder_test.go:223]: :rofl:

**Review by MrAlias — APPROVED** [review]

File: bridges/otelzap/encoder.go [hunk 3]
```diff
@@ -158,14 +161,20 @@ type arrayEncoder struct {
...
 }
 
-// TODO.
 func (a *arrayEncoder) AppendArray(v zapcore.ArrayMarshaler) error {
-	return nil
+	// TODO: Use arrayEncoder from a pool.
+	arr := &arrayEncoder{}
+	err := v.MarshalLogArray(arr)
+	a.elems = append(a.elems, log.SliceValue(arr.elems...))
+	return err
```
Comment (pellared) [inline bridges/otelzap/encoder.go:169]: Can you add a test that covers this scenario?

Comment (khushijain21) [reply bridges/otelzap/encoder.go:169]: It is covered by AddArray (nested) case

Comment (pellared) [reply bridges/otelzap/encoder.go:169]: This is not true. Run the tests with a debugger. Only `objectEncoder.AddArray`  is covered.

**Review by pellared — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Pull request merged**